Sources/LeafKit/LeafTag.swift

```diff
@@ -0,0 +1,40 @@
+public protocol LeafTag {
+    func render(_ ctx: LeafContext) throws -> LeafData
+}
+
+public var defaultTags: [String: LeafTag] = [
+    "lowercased": Lowercased(),
+    "index": LoopIndex(),
+    "isFirst": LoopIsFirst(),
```

Comment:
I'm not sure I understand why these tags exist? Can't you just access the data directly from the context:

```leaf
#if(isFirst):

#endif
```

That should work even without the `isFirst` tag.

Reply:
I would definitely think that's the case but it doesn't as things stand, and checking into what proper behavior should be as I didn't really look too closely at this PR before submitting it - just ran tests and assumed all was good.

I hadn't seriously started to look into the lex/parse stage until the last couple of days, and currently the problem is that mutating lex "tags" into syntax "tags" of greater specificity is relying on the Syntax methods themselves from what I can tell.

EG, at the token stage, the loop item, the loop array, and the loop indices are all still `tag`s and the Loop Syntax is converting its parameters to `variables`, but these remain `custom` tags and won't reference the context data directly.

Reply:
I *think* the solution is to make LeafTags somewhat more robust in terms of being able to self-report to Parser what they expect or can handle, and what they may publish. For things like Loop, that would make it much easier to allow varied syntax so that observable properties of a tag can function more like Swift - eg, a loop could be written as

```#for((i, name) in names):
    #(i) - #(name)
    #if(index == names.count): #(i) is last element#endif
```

Reply:
Yknow what I just checked this again since it seemed weird, and I'm wrong. For some reason the other day it was failing for me without the Tags, but I suspect it may have been a mangled build directory on my main computer because a clean build without them is working fine now. Strange.